py/picca/delta_extraction/utils_pk1d.py

```diff
@@ -213,44 +214,105 @@ def spectral_resolution(wdisp,
 
     return reso
 
+@njit#("f8[:](float32[:, :], i8)")
+def _find_nonzero_abs_min_per_row(reso_matrix, num_rows):
+    """Find the non-zero absolute minimum of the resolution matrix
+    for each row
+
+    Arguments
+    ---------
+    reso_matrix: 2D array of shape (num_diags, num_rows)
+    Resolution matrix
+
+    num_rows: int
+    Number of rows
+
+    Return
+    ------
+    nonzero_abs_min_per_row: 1D array
+    An array of size num_rows with non-zero abs. min
+    """
+    nonzero_abs_min_per_row = np.zeros(num_rows)
+    for irow in range(num_rows):
+        row = reso_matrix[:, irow]
+        nonzero_idx = row.nonzero()[0]
+        if nonzero_idx.size == 0:
+            continue
+
+        nonzero_abs_min_per_row[irow] = np.abs(row[nonzero_idx]).min()
+
+    return nonzero_abs_min_per_row
 
 def spectral_resolution_desi(reso_matrix, lambda_):
     """Compute the spectral resolution for DESI spectra
 
     Arguments
     ---------
-    reso_matrix: array
+    reso_matrix: 2D float32 array of shape (num_diags, num_rows)
     Resolution matrix
 
-    lambda_: array or None
-    Logarithm of the wavelength (in Angstroms)
+    lambda_: 1D array
+    Wavelength (in Angstroms)
 
     Return
     ------
     reso_in_km_per_s: array
     The spectral resolution
     """
-    delta_log_lambda = np.diff(np.log10(lambda_))
+    if lambda_ is None:
+        return None
+
+    delta_log_lambda = np.empty_like(lambda_)
+    delta_log_lambda[:-1] = np.diff(np.log10(lambda_))
     #note that this would be the same result as before (except for the missing bug) in
     #case of log-uniform binning, but for linear binning pixel size chenges wrt lambda
-    delta_log_lambda = np.append(
-        delta_log_lambda,
-        [delta_log_lambda[-1] + (delta_log_lambda[-1] - delta_log_lambda[-2])])
-    reso = np.clip(reso_matrix, 1.0e-6, 1.0e6)
+    delta_log_lambda[-1] = delta_log_lambda[-2] + (delta_log_lambda[-2] - delta_log_lambda[-3])
+
+    num_diags, num_rows = reso_matrix.shape
+    num_offdiags = num_diags // 2
+    # num_offdiags = reso_matrix.shape[0] // 2
+
+    # shift every row to a small positive value
+    # new resolution model is then Gaussian + constant
+    # using an arbitrary epsilon is not stable
+    # use nonzero absolute minimum of the row
+    nonzero_abs_min_per_row = _find_nonzero_abs_min_per_row(reso_matrix, num_rows)
+
+    # mask out rows that are all zeros
+    w = nonzero_abs_min_per_row>0
+    if not np.any(w):
+        return np.zeros_like(lambda_), np.zeros_like(lambda_)
+
+    shift = reso_matrix.min(axis=0) - nonzero_abs_min_per_row
```

Comment:
so this is to shift up the reso_matrix elements in case there are negaftive/zero ones? To avoid NANs later in the process? I guess it's never strongly negative and this is fine and just catching a numerical effect.
Or is it becoming strongly negative? In which case a more sophisticated solution might be necessary...